File: src/main/java/net/Chidoziealways/everythingjapanese/structure/ModProcessorLists.java
package net.Chidoziealways.everythingjapanese.structure;

import com.google.common.collect.ImmutableList;
import net.Chidoziealways.everythingjapanese.EverythingJapanese;
import net.minecraft.core.HolderGetter;
import net.minecraft.core.registries.Registries;
import net.minecraft.data.worldgen.BootstrapContext;
import net.minecraft.resources.ResourceKey;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.tags.BlockTags;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.levelgen.structure.templatesystem.*;

import java.util.List;

public class ModProcessorLists {
    private static final ResourceKey<StructureProcessorList> EMPTY = createKey("empty");
    public static final ResourceKey<StructureProcessorList> HELL_TEMPLE_START_DEGRADATION = createKey("hell_temple_start_degradation");

    private static ResourceKey<StructureProcessorList> createKey(String pName) {
        return ResourceKey.create(Registries.PROCESSOR_LIST, ResourceLocation.fromNamespaceAndPath(EverythingJapanese.MOD_ID, pName));
    }

    private static void register(
            BootstrapContext<StructureProcessorList> pContext, ResourceKey<StructureProcessorList> pKey, List<StructureProcessor> pProcessors
    ) {
        pContext.register(pKey, new StructureProcessorList(pProcessors));
    }

    public static void bootstrap(BootstrapContext<StructureProcessorList> pContext) {
        HolderGetter<Block> holdergetter = pContext.lookup(Registries.BLOCK);
        ProcessorRule processorrule = new ProcessorRule(new RandomBlockMatchTest(Blocks.BLACKSTONE, 0.01F), AlwaysTrueTest.INSTANCE, Blocks.GILDED_BLACKSTONE.defaultBlockState());
        ProcessorRule processorrule1 = new ProcessorRule(new RandomBlockMatchTest(Blocks.GILDED_BLACKSTONE, 0.5F), AlwaysTrueTest.INSTANCE, Blocks.BLACKSTONE.defaultBlockState());
        register(pContext, EMPTY, ImmutableList.of());

        register(
                pContext,
                HELL_TEMPLE_START_DEGRADATION,
                ImmutableList.of(
                        new RuleProcessor(
                                ImmutableList.of(
                                        new ProcessorRule(new RandomBlockMatchTest(Blocks.DEEPSLATE_BRICKS, 0.3F), AlwaysTrueTest.INSTANCE, Blocks.CRACKED_DEEPSLATE_BRICKS.defaultBlockState()),
                                        new ProcessorRule(new RandomBlockMatchTest(Blocks.DEEPSLATE_TILES, 0.3F), AlwaysTrueTest.INSTANCE, Blocks.CRACKED_DEEPSLATE_TILES.defaultBlockState()),
                                        new ProcessorRule(new RandomBlockMatchTest(Blocks.SOUL_LANTERN, 0.05F), AlwaysTrueTest.INSTANCE, Blocks.AIR.defaultBlockState())
                                )
                        ),
                        new ProtectedBlockProcessor(BlockTags.FEATURES_CANNOT_REPLACE)
                )
        );
    }
}

File: src/main/kotlin/net/Chidoziealways/everythingjapanese/structure/ModProcessorLists.kt
package net.Chidoziealways.everythingjapanese.structure

import com.google.common.collect.ImmutableList
import net.Chidoziealways.everythingjapanese.JAPANESE_MOD_ID
import net.minecraft.core.registries.Registries
import net.minecraft.data.worldgen.BootstrapContext
import net.minecraft.resources.ResourceKey
import net.minecraft.resources.Identifier
import net.minecraft.tags.BlockTags
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.level.levelgen.structure.templatesystem.*

object ModProcessorLists {
    private val EMPTY: ResourceKey<StructureProcessorList> = createKey("empty")
    val HELL_TEMPLE_START_DEGRADATION: ResourceKey<StructureProcessorList> = createKey("hell_temple_start_degradation")
    val HELL_TEMPLE_COURT_DEGRADATION: ResourceKey<StructureProcessorList> = createKey("hell_temple_court_degradation")

    val SHINTO_SHRINE_DEGRADATION = createKey("shinto_shrine_torii_degradation")

    private fun createKey(pName: String): ResourceKey<StructureProcessorList> {
        return ResourceKey.create(
            Registries.PROCESSOR_LIST,
            Identifier.fromNamespaceAndPath(JAPANESE_MOD_ID, pName)
        )
    }

    private fun register(
        pContext: BootstrapContext<StructureProcessorList>,
        pKey: ResourceKey<StructureProcessorList>,
        pProcessors: MutableList<StructureProcessor>
    ) {
        pContext.register(pKey, StructureProcessorList(pProcessors))
    }

    fun bootstrap(pContext: BootstrapContext<StructureProcessorList>) {
        val holdergetter = pContext.lookup(Registries.BLOCK)
        val processorrule = ProcessorRule(
            RandomBlockMatchTest(Blocks.BLACKSTONE, 0.01f),
            AlwaysTrueTest.INSTANCE,
            Blocks.GILDED_BLACKSTONE.defaultBlockState()
        )
        val processorrule1 = ProcessorRule(
            RandomBlockMatchTest(Blocks.GILDED_BLACKSTONE, 0.5f),
            AlwaysTrueTest.INSTANCE,
            Blocks.BLACKSTONE.defaultBlockState()
        )
        register(pContext, EMPTY, ImmutableList.of<StructureProcessor>())

        register(
            pContext,
            SHINTO_SHRINE_DEGRADATION,
            ImmutableList.of(
                RuleProcessor(
                    ImmutableList.of(
                        ProcessorRule(
                            RandomBlockMatchTest(Blocks.STONE, 0.4f),
                            AlwaysTrueTest.INSTANCE,
                            Blocks.COBBLESTONE.defaultBlockState()
                        ),
                        ProcessorRule(
                            RandomBlockMatchTest(Blocks.STONE_BRICKS, 0.2f),
                            AlwaysTrueTest.INSTANCE,
                            Blocks.CRACKED_STONE_BRICKS.defaultBlockState()
                        )
                    )
                ),
                ProtectedBlockProcessor(BlockTags.FEATURES_CANNOT_REPLACE)
            )
        )

        register(
            pContext,
            HELL_TEMPLE_COURT_DEGRADATION,
            ImmutableList.of<StructureProcessor>(
                RuleProcessor(
                    ImmutableList.of( // ⚒️ Cracked Deepslate Tiles (30% chance)
                        ProcessorRule(
                            RandomBlockMatchTest(Blocks.DEEPSLATE_TILES, 0.3f),
                            AlwaysTrueTest.INSTANCE,
                            Blocks.CRACKED_DEEPSLATE_TILES.defaultBlockState()
                        ),  // 🧱 Remove bricks (10% chance)
                        ProcessorRule(
                            RandomBlockMatchTest(Blocks.DEEPSLATE_BRICKS, 0.1f),
                            AlwaysTrueTest.INSTANCE,
                            Blocks.AIR.defaultBlockState()
                        ),  // 🌿 Turn bricks mossy (20%)
                        ProcessorRule(
                            RandomBlockMatchTest(Blocks.POLISHED_BLACKSTONE_BRICKS, 0.2f),
                            AlwaysTrueTest.INSTANCE,
                            Blocks.MOSSY_COBBLESTONE.defaultBlockState()
                        ),  // 🔥 Basalt corruption (random blackstone)
                        ProcessorRule(
                            RandomBlockMatchTest(Blocks.BASALT, 0.25f),
                            AlwaysTrueTest.INSTANCE,
                            Blocks.BLACKSTONE.defaultBlockState()
                        ),  // 💀 Curse with soul sand (5% floor tiles)
                        ProcessorRule(
                            RandomBlockMatchTest(Blocks.STONE_BRICKS, 0.05f),
                            AlwaysTrueTest.INSTANCE,
                            Blocks.SOUL_SAND.defaultBlockState()
                        ),  // 🕸️ Add cobwebs for spice (only 5%)
                        ProcessorRule(
                            RandomBlockMatchTest(Blocks.AIR, 0.05f),
                            AlwaysTrueTest.INSTANCE,
                            Blocks.COBWEB.defaultBlockState()
                        )
                    )
                ),
                ProtectedBlockProcessor(BlockTags.FEATURES_CANNOT_REPLACE)
            )
        )

        register(
            pContext,
            HELL_TEMPLE_START_DEGRADATION,
            ImmutableList.of(
                RuleProcessor(
                    ImmutableList.of(
                        ProcessorRule(
                            RandomBlockMatchTest(Blocks.DEEPSLATE_BRICKS, 0.3f),
                            AlwaysTrueTest.INSTANCE,
                            Blocks.CRACKED_DEEPSLATE_BRICKS.defaultBlockState()
                        ),
                        ProcessorRule(
                            RandomBlockMatchTest(Blocks.DEEPSLATE_TILES, 0.3f),
                            AlwaysTrueTest.INSTANCE,
                            Blocks.CRACKED_DEEPSLATE_TILES.defaultBlockState()
                        ),
                        ProcessorRule(
                            RandomBlockMatchTest(Blocks.SOUL_LANTERN, 0.05f),
                            AlwaysTrueTest.INSTANCE,
                            Blocks.AIR.defaultBlockState()
                        )
                    )
                ),
                ProtectedBlockProcessor(BlockTags.FEATURES_CANNOT_REPLACE)
            )
        )
    }
}
